File: test_cli.py
import os, sys

if len(sys.argv) < 2:
    print("Usage: %s <cmd>" % sys.argv[0])
    print("Available commands:")
    print("TEMPERATURE - get the temperature in Celcius")
    exit(1);

fifo_client = "/var/run/imsg/rx_pipe"
fifo_server = "/var/run/thermo_lm75/rx_pipe"

if not os.path.exists(fifo_client):
    os.mkfifo(fifo_client)

pipeout = os.open(fifo_server, os.O_WRONLY)
if pipeout == None:
    print("can't open %s" % fifo_server)
    exit(1)

n = os.write(pipeout, sys.argv[1]+"\n")

if n != len(sys.argv[1] + "\n"):
    print("error writing to %s" % fifo_server)
    exit(1)

pipein = open(fifo_client, 'r')
if pipein == None:
    print("can't open %s" % fifo_client);
    exit(1)

line = pipein.readline()
print(line),

os.unlink(fifo_client)

File: test_cli.c
#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>

static const char *fifo_client = "/var/run/imsg/rx_pipe";
static const char *fifo_server = "/var/run/thermo_lm75/rx_pipe";
static int serverfd, clientfd;

#define error(fmt, ...) \
    do { fprintf(stderr, "errno(%s): " fmt "\n", strerror(errno), ##__VA_ARGS__); exit(1); } \
    while(0)

static void removefifo()
{
    unlink(fifo_client);
}

int main(int argc, char **argv)
{
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <request>\n", argv[0]);
        exit(1);
    }

    unsigned char buf[64];
    FILE *fclient;

    if (mkfifo(fifo_client, S_IRUSR | S_IWUSR | S_IWGRP) < 0 && errno != EEXIST)
        error("mkfifo %s", fifo_client);

    if (atexit(removefifo) != 0)
        error("atexit");

    if ((serverfd = open(fifo_server, O_WRONLY)) < 0)
        error("open %s", fifo_server);

    snprintf(buf, sizeof(buf), "%s\n", argv[1]);

    if (write(serverfd, buf, strlen(buf)) != strlen(buf))
        error("write serverfd(%d)", serverfd);

    if ((clientfd = open(fifo_client, O_RDONLY)) < 0)
        error("open %s", fifo_client);

    if ((fclient = fdopen(clientfd, "r")) == NULL)
        error("fdopen %s failed", fifo_client);

    if (fgets(buf, sizeof(buf), fclient) == NULL)
        error("fgets failed");

    fputs(buf, stdout);
}
